[Patch 1/2] implement AP capability report test (4.6.1)

CoralMalachi opened the pull request:
## Description

According to the Wi-Fi CERTIFIED EasyMesh™ Test Plan v1.0, test 4.6.1, which is the AP capability report test, the MAUT is triggered by the CTT controller with an AP Capability Query message, and the MAUT should reply with AP Capability Report message.

According to Multi-AP-Spesification, when a Multi-AP Agent receives an AP
Capability Query message, it shall respond within one second with an AP
Capability Report message which includes one AP Capability TLV, and one AP Radio Basic Capabilities TLV for each AP radio.

Currently, the handling of the AP Capability Query message is in the responsibility of the son slave thread, which makes no sense, since one slave has no way to know the AP capabilities of the other slaves (and should not).

The handling of the AP Capability Query message should be part of the backhaul manager, and for that reason, the following changes were made:

- Each one of the radios sends its supported channels list as part of the cACTION_BACKHAUL_ENABLE
VS CMDU to the backhaul manager.

- The backhaul manager saves for each radio its supported channels list

- When the AP Capability Query message is received in the backhaul manager, an AP Capability Report message is created, which contains one AP Capability TLV, and one AP Radio Basic Capabilities TLV for each AP radio.


## Testbed status

_Those changes were checked on the testbed with the excepted results._

Commit:
tlvf: update ACTION_BACKHAUL_ENABLE VS CMDU

Preparative commit.

Update the YAML file of  cACTION_BACKHAUL_ENABLE to contain a known
length list of the supported channels.

Remove ruid field from cACTION_BACKHAUL_ENABLE.
ruid is now radio_mac so field ruid is no longer needed.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>
Signed-off-by: Mario Maz <mmazzapater@gmail.com>

Commit:
agent: send supported channels list to backhaul manager

Preparative commit.

Send the supported channels list as part of the cACTION_BACKHAUL_ENABLE
VS CMDU.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
agent: add a map to save capabilies of each radio

Preparative commit.

Add an unordered map to the backhaul manager with a key of type
sMacAddr, which represents the radio identifier, and a vector of
supported channels by this radio as a value.

Currently there is already another map containing radio information,
which is also indexed by radio identifier. So create a new struct
called sRadioInfo containing VAPs list and supported channels (to be
used as map value) and use a single map of radio information for
cleanliness.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>
Signed-off-by: Mario Maz <mmazzapater@gmail.com>

Commit:
agent: save the list of supported channels

Preparative commit.

When the backhaul manager receives a cACTION_BACKHAUL_ENABLE VS CMDU
from one of the slaves, save the supported channels list info in the
backhaul manager database, so we can use it later (this will be
explained in later commits)

Signed-off-by: Coral Malachi <coral.malachi@intel.com>
Signed-off-by: Mario Maz <mmazzapater@gmail.com>

Commit:
agent: backhaul manager: add ap capability handler

According to Multi-AP-Specification, when a Multi-AP Agent receives an AP
Capability Query message, it shall respond within one second with an AP
Capability Report message and include one AP Radio Basic Capabilities TLV
for each AP radio.

For that reason, the handle method of the AP Capability Query message
should be part of the backhaul manager instead of the son slave.

To be able to create an AP Radio Basic Capabilities TLV for each AP radio,
the backhaul manager needs to save this info for each radio, and this is
the reason for creating a map containing supported channels which has the
information we need for each radio.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>
Signed-off-by: Mario Maz <mmazzapater@gmail.com>

Commit:
agent: son_slave: remove ap capability handle method

Now, after moving the responsibility of handling the AP Capability Query
Message from the son slave to the backhaul manager, remove
slave_thread::handle_ap_capability_query method, which is unused anymore.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
test flows: update ap capabilities test

Now, after moving the responsibility of handling the AP Capability Query
Message from the son slave to the backhaul manager, the
"AP_CAPABILITY_QUERY_MESSAGE" log message will be found in the
beerocks_agent.log.

Therefore, update the ap capability query test to search for this log
in the correct file.

Check also that controller receives the response message.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>
Signed-off-by: Mario Maz <mmazzapater@gmail.com>

Commit:
tlvf: fix a typo error

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
tlvf: update tlvf_conf

Preparative commit.

Add the following TLVs to the `tlvf_conf` YAML file so the TLVF will
auto-generate their source files. Those TLVS will be needed in the
second part of the test 4.6.2 implementation.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
tlvf: add TLVs auto-generated files

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

tomereli commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -3091,6 +3091,12 @@ bool slave_thread::slave_fsm(bool &call_slave_select)
+        std::copy_n(hostap_params.supported_channels, message::SUPPORTED_CHANNELS_LENGTH,
+                    &std::get<1>(tuple_supported_channels));
+
+        bh_enable->ruid() = network_utils::mac_from_string(config.radio_identifier);
```
Important - see #691 - this will need to change to the radio mac if #431 will be merged before this change (and it probably will)

arnout replied:
As mentioned, this will be removed.

Also, please always assign values in the same order as they are specified in the yaml file. This is essential in case there's a dynamic length list before it. It's not the case now, but it might be in the future.

mariomaz replied:
I've removed the ruid field

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1491,7 +1491,18 @@ bool backhaul_manager::handle_slave_backhaul_message(std::shared_ptr<SSlaveSocke
+        auto channels                 = &std::get<1>(tuple_supported_channels);
+
+        std::vector<beerocks::message::sWifiChannel> supported_channels_vec;
+        for (uint8_t i = 0; i < beerocks::message::SUPPORTED_CHANNELS_LENGTH; i++) {
```
use `int` please

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1759,6 +1764,83 @@ bool backhaul_manager::handle_1905_1_message(ieee1905_1::CmduMessageRx &cmdu_rx,
+        std::vector<uint8_t> operating_classes;
+        auto ruid = radio_cap.first;
+
+        auto radio_basic_caps = cmdu_tx.addClass<wfa_map::tlvApRadioBasicCapabilities>();
```
This is duplicated code - please use `add_radio_basic_capabilities` and refactor it accordingly if needed so you can use it from here.

arnout replied:
That will be difficult since both of them use private members (from slave_thread and backhaul_manager). So the best we can do is to add a function (where?) that fills in the capabilities based on a channel list.

Or maybe we should add a full-fledged class for the radio description (radio UID, channels, associated clients, and all the other info we need to cache) and use it in both backhaul manager and slave.

CoralMalachi replied:
we also can pass those private members as arguments to this function.

mariomaz replied:
I've created a new class which for the moment, contains just one static method: add_ap_radio_basic_capabilities. Required private members are passed as parameters.
This class is intended to contain code to build up TLVs and to be shared between backhaul_manager_thread and son_slave_thread (or any other agent thread)

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1759,6 +1764,83 @@ bool backhaul_manager::handle_1905_1_message(ieee1905_1::CmduMessageRx &cmdu_rx,
+        }
+
+        radio_basic_caps->radio_uid() = ruid;
+        //TODO get maximum supported VAPs from DWPAL
```
Also change the comment (in `add_radio_basic_capabilities()`) - to // TODO get maximum supported VAPs from AP Capabilities from BWL.

mariomaz replied:
Done

morantr commented:
commit message headline spelling mistake 418e1dc0eba335ab740ea3af76577ecb84dcd5e4
eahc -> each

arnout commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_backhaul.yaml:
```diff
@@ -58,6 +58,10 @@ cACTION_BACKHAUL_ENABLE:
+  ruid: sMacAddr
+  supported_channels_list:
+    _type: beerocks::message::sWifiChannel 
+    _length: [ "beerocks::message::SUPPORTED_CHANNELS_LENGTH" ]
```
I don't particularly like these fixed-length lists. Why not add make it a dynamic length list by adding a supported_channels_list_len before it?

CoralMalachi replied:
I created a fixed-length size list since I saw that in the other YAML files this supported_channels is defined as a fixed-length size list (e.g. beerocks_message_apmanager.yaml, beerocks_message_common.yaml ).
 I agree with you that it makes more sense to make it a dynamic length list, but shouldn't we do it in the rest of the place we declare this list?

mariomaz replied:
Looks like `supported_channels` is treated as a fixed-length list all over the code. IMO it should not be defined as a dynamic length list.

arnout commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_backhaul.yaml:
```diff
@@ -58,6 +58,10 @@ cACTION_BACKHAUL_ENABLE:
   wireless_iface_type: uint8_t  
   mem_only_psk: uint8_t
   backhaul_preferred_radio_band: uint8_t
+  ruid: sMacAddr
```
Now the ruid==mac change has been merged, this is no longer needed. radio_mac is available in the action header.

mariomaz replied:
Done

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -3091,6 +3091,12 @@ bool slave_thread::slave_fsm(bool &call_slave_select)
                                   config.backhaul_wireless_iface.c_str(),
                                   message::IFACE_NAME_LENGTH);
 
+        auto tuple_supported_channels = bh_enable->supported_channels_list(0);
```
You have to check the "return value":
```
if (!std::get<0>(tuple_supported_channels)) {
    LOG(ERROR) << ...;
    do error handline
}
```

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h:
```diff
@@ -230,6 +230,8 @@ class backhaul_manager : public btl::transport_socket_thread {
     // vaps map of ruid (key) to VapsList (value)
     std::unordered_map<sMacAddr, beerocks_message::sVapsList> m_vaps_map;
+    std::unordered_map<sMacAddr, std::vector<beerocks::message::sWifiChannel>>
+        m_radio_capabilities_map;
```
I think these two should be merged into a single map with a struct radio_info (that still has to be defined) as the rhs.

mariomaz replied:
LGTM
Is there any reason to use both operator[] and insert() to add items to the maps?
insert() does not insert if key already exists.

mariomaz replied:
Done

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1491,7 +1491,18 @@ bool backhaul_manager::handle_slave_backhaul_message(std::shared_ptr<SSlaveSocke
+        auto channels                 = &std::get<1>(tuple_supported_channels);
+
+        std::vector<beerocks::message::sWifiChannel> supported_channels_vec;
+        for (uint8_t i = 0; i < beerocks::message::SUPPORTED_CHANNELS_LENGTH; i++) {
```
See, this is annoying: we're creating a std::vector that is always the same size! Dynamic length would be better.

BTW also this is a container where the size doesn't change and it's known up-front (even in case of a dynamic length list) so use a std::array instead.

It would actually be nice if we would add a tlvf function that would return such an array for a dynamic list. It would simplify our life a lot in many places.

tomereli replied:
@LiorAmram Can you add such a function?

arnout reviewed (commented):
Partial review only - I got distracted by other stuff.

I thought it would have been nicer if the slaves would create the EasyMesh TLVs and send those to the backhaul manager, which collects them in a single CMDU. Unfortunately, the way the TLV factory is built, it's quite complicated to take that approach. So this approach of storing all info on both backhaul manager and slave is OK.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1759,6 +1764,83 @@ bool backhaul_manager::handle_1905_1_message(ieee1905_1::CmduMessageRx &cmdu_rx,
+                                                  const std::string &src_mac)
+{
+    //TODO - this is a stub handler for the purpose of controller certification testing,
+    //       will be implemented later on agent certification
```
This TODO is no longer valid.

mariomaz replied:
Removed

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1759,6 +1764,83 @@ bool backhaul_manager::handle_1905_1_message(ieee1905_1::CmduMessageRx &cmdu_rx,
+        LOG(ERROR) << "addClass wfa_map::tlvApCapability has failed";
+        return false;
+    }
+
```
We should add a comment here that the capability bitmask stays at 0 because we don't support unassociated STA metrics or agent-initiated steering.

mariomaz replied:
Done

mariomaz committed:
tlvf: update autogen files

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

arnout commented on tools/docker/tests/test_flows.sh:
```diff
@@ -198,7 +198,7 @@ test_ap_capability_query() {
     check send_CAPI_command gateway "DEV_SEND_1905,DestALid,$mac_agent1,MessageTypeValue,0x8001" $redirect
     sleep 1
     dbg "Confirming ap capability query has been received on agent"
-    check docker exec repeater1 sh -c 'grep -i -q "AP_CAPABILITY_QUERY_MESSAGE" /tmp/$USER/beerocks/logs/beerocks_agent_wlan0.log'
```
Please also update the controller to log the details of the AP Capability Report it receives, and check that in test_flows.

mariomaz replied:
Sorry, but I need a detailed explanation on how to do this

mariomaz replied:
No, I didn't

mariomaz replied:
Added handler in controller and check in test_flows.sh

arnout reviewed (changes requested):
The changes that I'm asking for are so invasive that a re-review will be required, so marking as changes requested.

mariomaz committed:
bcl: add const type modifier to supported_channels parameter

Add `const` type modifier to supported_channels parameter, which is not
being modified inside any of the methods.

This is a preparatory commit so we can later invoke these methods using
std::array::data, which is const.

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
agent: create TLVF utility functions for agent

Some TLV management functions are required both from
backhaul_manager_thread and from son_slave_thread.

Create a file with a code to be shared among different agent threads
related to TLV creation.

This is a preparatory commit. Following commits will remove duplicated code
and call the methods defined in this file.

Currently this file contains one class with one method only:
add_ap_radio_basic_capabilities. It is expected to be completed with other
methods in the future, as necessities arise.

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
controller: add handler for AP Capabilities Report message

Current message handler does nothing with received message but
logging it (for testing purposes with test_flows.sh)

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

tomereli commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_backhaul.yaml:
```diff
@@ -58,7 +58,6 @@ cACTION_BACKHAUL_ENABLE:
   wireless_iface_type: uint8_t  
   mem_only_psk: uint8_t
   backhaul_preferred_radio_band: uint8_t
-  ruid: sMacAddr
```
This should be a fixup commit to the first one.

mariomaz replied:
Fixed

tomereli commented on common/beerocks/tlvf/AutoGenerated/include/beerocks/tlvf/beerocks_message_backhaul.h:
```diff
@@ -147,7 +147,6 @@ class cACTION_BACKHAUL_ENABLE : public BaseClass
         uint8_t& wireless_iface_type();
         uint8_t& mem_only_psk();
         uint8_t& backhaul_preferred_radio_band();
-        sMacAddr& ruid();
```
Same for this commit - should be squashed to the original one in this series.

mariomaz replied:
Fixed

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1504,7 +1504,7 @@ bool backhaul_manager::handle_slave_backhaul_message(std::shared_ptr<SSlaveSocke
         }
 
-        m_radio_capabilities_map.insert({request->ruid(), supported_channels_vec});
+        m_radio_capabilities_map.insert({request->iface_mac(), supported_channels_vec});
```
Same here - another fixup commit.

mariomaz replied:
Fixed

tomereli commented on agent/src/beerocks/slave/agent_tlvf_utils.cpp:
```diff
@@ -0,0 +1,70 @@
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
```
This commit's purpose is to remove duplicate code which is great. The commit message is also great.
It's hilarious that the commit message itslef is duplicated twice though 🤣

mariomaz replied:
Fixed

tomereli commented on agent/src/beerocks/slave/agent_tlvf_utils.cpp:
```diff
@@ -0,0 +1,70 @@
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
+ *
+ * Copyright (c) 2020 Intel Corporation
```
File name should be tlvf_utils.cpp (its in the agent's path so its clear it is for the agent).
Also, since you created this file - you can also add a Copyright of you and your company.

mariomaz replied:
Renamed. 
I think I don't deserve any credit for this code since I haven't created it, just brought it from another file as is.

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1788,53 +1788,13 @@ bool backhaul_manager::handle_ap_capability_query(ieee1905_1::CmduMessageRx &cmd
-                    idx++;
-                }
-            }
+        std::array<beerocks::message::sWifiChannel, beerocks::message::SUPPORTED_CHANNELS_LENGTH>
```
I like this!

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1784,6 +1784,9 @@ bool backhaul_manager::handle_ap_capability_query(ieee1905_1::CmduMessageRx &cmd
         return false;
     }
 
+    // Capability bitmask is set to 0 because neither unassociated STA link metrics
```
I guess this commit should also be a fixup

mariomaz replied:
Fixed

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1788,7 +1788,6 @@ bool backhaul_manager::handle_ap_capability_query(ieee1905_1::CmduMessageRx &cmd
     // reporting or agent-initiated RCPI-based steering are supported
 
     for (const auto &radio_cap : m_radio_capabilities_map) {
-        std::vector<uint8_t> operating_classes;
```
Also a fixup

mariomaz replied:
Fixed

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h:
```diff
@@ -230,10 +230,13 @@ class backhaul_manager : public btl::transport_socket_thread {
-    std::unordered_map<sMacAddr, std::vector<beerocks::message::sWifiChannel>>
-        m_radio_capabilities_map;
+
+    struct SRadioInfo {
```
```suggestion
    struct sRadioInfo {
```

mariomaz replied:
Fixed

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h:
```diff
@@ -230,10 +230,13 @@ class backhaul_manager : public btl::transport_socket_thread {
-    std::unordered_map<sMacAddr, std::vector<beerocks::message::sWifiChannel>>
-        m_radio_capabilities_map;
+
+    struct SRadioInfo {
```
Please add doxygen comment on this new structure

mariomaz replied:
Fixed

tomereli commented on tests/test_flows.sh:
```diff
@@ -203,8 +203,13 @@ test_ap_capability_query() {
     check_error=0
     check send_CAPI_command ${GATEWAY} "DEV_SEND_1905,DestALid,$mac_agent1,MessageTypeValue,0x8001" $redirect
     sleep 1
+    
```
Should be a fixup to the commit which added the test and just reorder it to be last

mariomaz replied:
Fixed

tomereli reviewed (approved):
Most my comments are regarding squashing / fixups - but they are important, so please do your best to organize and squash commits so the whole change can be "digested" by a first time reviewer.
If this is too much work to do it all - then do what you can.
Other than that and some more small comments, LGTM - I think after some cleanups it should be good for merge, so approving.

mariomaz committed:
agent: remove duplicated code creating AP Radio Basic Capabilities TLV

Code to create AP Radio Basic Capabilities TLV is required from
backhaul_manager_thread and from son_slave_thread.

Shared code has been factored out to a shared file in previous commit
and thus can now be removed from son_slave_thread

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz commented:
Commits on PR "Feature/add associated clients tlv to topology response #786" where pushed on top of the commits of this PR and they have already been merged so this PR can be closed.
